### PR TITLE
add mkDocs

### PR DESCRIPTION
**Pull request by Geometer1729**
~~I'm trying to add `,docs` like we have `,lint` and `,format`. I'm using `nix build .#haddock` which I assume is bad, I'm not sure how to avoid that.~~
Adds `lib.mkDocs` which can be used like:
https://github.com/mlabs-haskell/plutus-simple-model/blob/3d9b23e4ffbd0718661ad1435a422ba90b6d8912/flake.nix#L37-L42
to add `nix run .#docs`

### CODE REVIEW

**Comment by AriFordsham** [comment]
@Geometer1729 Have you tested this in a pure nix shell to ensure you have all the python dependencies it needs?

**Comment by AriFordsham** [comment]
Yup, We probably want to wait, at least a bit, To see if we can get this merged upstream.

**Comment by Geometer1729** [comment]
> @Geometer1729 Have you tested this in a pure nix shell to ensure you have all the python dependencies it needs?

I added `which python3` to the script (temporarily) to make sure it was using python itself from the nix store but I didn't think of that no.

File: docs.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+export LC_CTYPE=C.UTF-8
+export LC_ALL=C.UTF-8
+export LANG=C.UTF-8
+
+echo generating docs
+# I assume this is bad
+nix build .#haddock
+echo serving docs
+python3 -m http.server --directory ./result/share/doc 8080
```
Comment (AriFordsham) [inline docs.sh:11]: `./result` only contains the result of your _last_ `nix build` (which may be a library build, not a haddock build, for example), which happens to be correct here, but there must be a more robust way of doing this.

File: flake.nix [hunk 2]
```diff
@@ -73,6 +73,15 @@
         }).outPath;
       };
 
+      mkDocs = pkgs: with pkgs; writeShellApplication {
+        name = ",docs";
+        runtimeInputs = [
+          pkgs.python3
```
Comment (AriFordsham) [inline flake.nix:79]: You don't _know_ that what you have here is everything needed (although I'd assume it does). Python is also a rather large dependency to pull in.

File: docs.sh (hunk 1 above)
Comment (brainrake) [reply docs.sh:11]: you can pass `${self.packages.${system}.haddock}` to the script and serve that

File: flake.nix (hunk 2 above)
Comment (Geometer1729) [reply flake.nix:79]: Fair enough, I checked it with `nix-shell -p python3` and it seems to work.

File: docs.sh (hunk 1 above)
Comment (Geometer1729) [reply docs.sh:11]: Self doesn't seem to have a packages attribute. It seems like I would need to get the output of the flake using mlabs-tooling ie. psm but I don't know how to do that.

**Comment by L-as** [comment]
Since this builds code, you should instead use `nix run .#serve-docs`, and use `writeShellScript` to create a script that serves the docs.

File: flake.nix [hunk 3]
```diff
@@ -40,6 +40,15 @@
     lib = {
       inherit moduleMod mkHackageMod;
 
+      mkDocs = target: pkgs: pkgs.writeShellScript "docs"
+        ''
+          set -xe
+          export LC_CTYPE=C.UTF-8
+          export LC_ALL=C.UTF-8
+          export LANG=C.UTF-8
+          ${pkgs.python3}/bin/python3 -m http.server --directory ${target}/share/doc 8080
```
Comment (L-as) [inline flake.nix:49]: Using webfs seems better than depending on Python

**Comment by aciceri** [comment]
As it's written now this PR just adds a function to `lib` that takes `target` and returns a derivation which builds a shell script that exposes it. It's not adding this script to `devShell` nor flake `apps` or I'm missing something?

Also I suggest to use `writeShellApplication` instead of `writeShellScript`, it prepends automatically
```sh
set -o errexit
set -o nounset
set -o pipefail
```
and checks the code against [shellcheck](https://www.shellcheck.net/)
Even if perhaps it's a bit overkill for a so small script...

Also personally I don't like overriding locales like this, maybe one set his system locale to Lojban and you would show him docs in an old, ambiguous and boring English, I would prefer something like:

```sh
export LANG=${$LANG:-C.UTF-8}
# ...for other values too...
```

Moreover there are more idiomatic way to manage this if we use `flake-parts`, both if we use `mission-control` (see #26, I would suggest it) and we don't. For example I don't like putting stuff that depends on `system` (this function would depend on `pkgs` that depends on `system`), for instance one could use overlays (or probably even a better `flake-parts` devilry).

**Comment by Geometer1729** [comment]
@aciceri 

> As it's written now this PR just adds a function to `lib` that takes `target` and returns a derivation which builds a shell script that exposes it. It's not adding this script to `devShell` nor flake `apps` or I'm missing something?
> Also I suggest to use `writeShellApplication` instead of `writeShellScript`, it prepends automatically

I should've updated my comment, @L-as  suggested I go to `writeShellScript` and make it `nix run .#serve-docs` rather than
a dev shell command.

> Also personally I don't like overriding locales like this, 

I just copied the beginning of the similar scripts assuming it fixes some bug. I don't feel strongly about it.
@L-as do you know why the other scripts have it and rather this should too?

**Comment by L-as** [comment]
Some have it because of buggy Haskell programs.

**Comment by Geometer1729** [comment]
Fair enough, sounds  like it shouldn't be needed here and removing it doesn't seem to break anything so I removed the locales.
As for the `writeShellScript` vs `writeShellApplication` which should I go with @L-as @aciceri?

**Comment by L-as** [comment]
writeShellApplication always IMO

**Comment by aciceri** [comment]
I agree with Las, as rule of thumb I would always use it unless you've specific reasons to not use it.

> I just copied the beginning of the similar scripts assuming it fixes some bug. I don't feel strongly about it.

Yes I remember seeing that someone had problems when LOCALE was unset (maybe on darwin?) so setting it could makes sense. This is also a a matter of tastes by the way, do we really want completely hermetic dev shells?

> Adds `lib.mkDocs` which *can be used* [...] to add `nix run .#docs`
Ahhh now I understand, but why don't you directly add it here in this repo?

By the way it's weeks that I'm collecting issues about `mlabs-tooling.nix`, I would really like spending some full days on it but I haven't find time yet. I think we can merge this PR as is for now since I guess this is something you really need/want, I really hope that we'll implement other open issues soon, please open issues about what you would like to have (even if are unsure, just open them so we can discuss, remember to ping me).

**Review by aciceri — APPROVED** [review]

**Comment by aciceri** [comment]
You must hate me (sorry :smile:) but three little notes:

- `name` is the name of the derivation to build a script that exposes docs, so something like `serve-docs` would be more meaningful
- `set -e` is useless since it's already included by [`writeShellApplication`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/trivial-builders.nix#L329)
- if you pass `runtimeInputs = [ pkgs.webfs ]` to `writeShellApplication` then you can just write `webfs` instead of `${pkgs.webfs}/bin/webfs` in your script

**Comment by Geometer1729** [comment]
> You must hate me (sorry smile) but three little notes:

Not at all happy to learn some more nix :)

**Pull request merged**